tests/test_middleware.py

```diff
@@ -349,6 +349,66 @@ def test_skip_download_request_url(settings):
     assert crawler.stats.get_stats().get("downloader/response_count", 0) == 1
 
 
+class ResponseUrlSpider(scrapy.Spider):
+    url = None
+
+    def start_requests(self):
+        yield Request(url=self.url, callback=self.parse)
+
+    def parse(self, response: DummyResponse, url: ResponseUrl):
+        return {
+            "response": response,
+            "url": url,
+        }
+
+
+@inlineCallbacks
+def test_skip_download_response_url(settings):
+    item, url, crawler = yield crawl_single_item(
+        ResponseUrlSpider, ProductHtml, settings
+    )
+    assert isinstance(item["response"], Response) is True
+    # Even if the spider marked the response with DummyResponse, the response
+    # is still needed since ResponseUrl depends on it.
+    assert isinstance(item["response"], DummyResponse) is False
```

Comment:
I am unsure about this, setting `response: DummyResponse` in the signature and getting `response: Response` feels wrong, even if a response is actually acquired to provide `ResponseUrl`.

Reply:
I believe this should be fine since it's noted down in the docstrings of the middleware ([code ref](https://github.com/scrapinghub/scrapy-poet/blob/master/scrapy_poet/downloadermiddlewares.py#L72-L78)):
> This method checks if the request is really needed and if its
        download could be skipped by trying to infer if a ``Response``
        is going to be used by the callback or a Page Input.
        If the ``Response`` can be ignored, a ``utils.DummyResponse`` object is
        returned on its place. This ``DummyResponse`` is linked to the original
        ``Request`` instance.

Requesting for a `DummyResponse` isn't actually guaranteed but provided if downloading the request can be avoided.

Reply:
OK.

Since setting `response: DummyResponse` is meant to signify that you are not interested in the response object, and we want to eventually allow users to skip the parameter itself from the signature, I guess it does not matter what it contains.